internal/nri/domain.go

```diff
@@ -0,0 +1,109 @@
+package nri
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/cri-o/cri-o/internal/log"
+	"github.com/sirupsen/logrus"
+
+	nri "github.com/containerd/nri/pkg/adaptation"
+)
+
+// Domain implements the functions the generic NRI interface needs to
+// deal with pods and containers from a particular containerd namespace.
+type Domain interface {
+	// GetName() returns the containerd namespace for this domain.
+	GetName() string
+
+	// ListPodSandboxes list all pods in this namespace.
+	ListPodSandboxes() []PodSandbox
+
+	// ListContainer list all containers in this namespace.
+	ListContainers() []Container
+
+	// GetPodSandbox returns the pod for the given ID.
+	GetPodSandbox(string) (PodSandbox, bool)
+
+	// GetContainer returns the container for the given ID.
+	GetContainer(string) (Container, bool)
+
+	// UpdateContainer applies an NRI container update request in the namespace.
+	UpdateContainer(context.Context, *nri.ContainerUpdate) error
+
+	// EvictContainer evicts the requested container in the namespace.
+	EvictContainer(context.Context, *nri.ContainerEviction) error
```

Comment:
What's the use case for eviction? Eviction is handled by the kubelet today.

Reply:
Hmm... eviction is probably a misnomer at the NRI level wrt. what this really does. Nothing is being evicted/removed here. EvictContainer stops the container. There are cases where an NRI plugin might need to stop a best-effort container and this interface is for that.